### PR TITLE
[SPMD] Allow mark_shard scalar tensors

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
Support mark_shard scalar tensors.

Test Plan:
python test/spmd/test_xla_sharding.py -v -k test_mark_shard_scalar

### CODE REVIEW

**Comment by alanwaketan** [comment]
Do we plan to support such use case in the future?

**Comment by JackCaoG** [comment]
what's the expected behavior through?

**Comment by alanwaketan** [comment]
> what's the expected behavior through?

This is for open discussion. I would suggest we don't error out but instead ignore the operation if the tensor is a scalar given scalars can only be replicated.

**Comment by jonb377** [comment]
> Do we plan to support such use case in the future?

The only possible sharding on a scalar is pure replication and any unsharded tensors are implicitly replicated, so I don't see a need from the expressibility of the APIs.

I would say it is worth supporting though, so that mark_sharding can act on all tensors without regard for their rank. Right now, a user would need to special-case for rank-0 tensors when iterating over a set of tensors, while we could hide this behind mark_sharding.

**Comment by alanwaketan** [comment]
Okay, I would go ahead to make mark_shard not error out on this operation.

File: test/spmd/test_xla_sharding.py [hunk 1]
```diff
@@ -1047,6 +1047,29 @@ def test_from_cpu_shards_global_shape(self):
...
+    self.assertEqual(len(shards), self.n_devices)
+    # all shards are REPLICATED.
+    for i, shard in enumerate(shards):
+      self.assertEqual(shard.data.device, torch.device('cpu'))
+      self.assertTrue(torch.allclose(shard.data, torch.tensor(1.0)))
+      self.assertIsInstance(shard.indices, type(Ellipsis))
+      self.assertEqual(shard.replica_id, i)
+
+    # It looks like mesh_shape attribute is never implemented.
+    with self.assertRaises(AttributeError):
```
Comment (alanwaketan) [inline test/spmd/test_xla_sharding.py:1070]: @yeounoh What's the deal here? Should we implement it?

**Review by jonb377 — APPROVED** [review]
LGTM! Thanks for resolving this @alanwaketan

**Comment by alanwaketan** [comment]
Thanks Jon for approving the change.

File: test/spmd/test_xla_sharding.py (hunk 1 above)
Comment (yeounoh) [reply test/spmd/test_xla_sharding.py:1084]: Let's keep it unimplemented for now, I will revisit when integrate with DTensor.

**Comment by yeounoh** [comment]
Awesome, thank you @alanwaketan

**Review by yeounoh — COMMENTED** [review]
LGTM

**Pull request merged**